services/docker/docker-size.service.js

```diff
@@ -73,30 +84,77 @@ export default class DockerSize extends BaseJsonService {
     })
   }
 
-  transform({ tag, sort, data }) {
+  transform({ tag, sort, data, arch }) {
+    let sizeFromSpecifiedArchitecture
+
     if (!tag && sort === 'date') {
       if (data.count === 0) {
         throw new NotFound({ prettyMessage: 'repository not found' })
       } else {
-        return { size: data.results[0].full_size }
+        const latestEntry = data.results[0]
+
+        // If no tag is specified, and sorting is by date, check if any of the returned images has an architecture matching the arch parameter supplied by the user.
+        // If yes, return the size of the image with this arch.
+        // If not, return the value of the `full_size` in the latestEntry from the response.
+        // For details see: https://github.com/badges/shields/issues/8238
+
+        Object.values(latestEntry.images).forEach(img => {
+          if (img.architecture === arch) {
+            sizeFromSpecifiedArchitecture = img.size
+          }
+        })
+
+        return {
+          size: sizeFromSpecifiedArchitecture || latestEntry.full_size,
+        }
       }
     } else if (!tag && sort === 'semver') {
-      const [matches, versions] = data.reduce(
-        ([m, v], d) => {
+      // If no tag is specified, and sorting is by semver, first filter out the entry containing the latest semver from the response with Docker images.
+      // Then check if any of the returned images for this entry has an architecture matching the arch parameter supplied by the user.
+      // If yes, return the size of the image with this arch.
+      // If not, return the value of the `full_size` from the entry matching the latest semver.
+
+      const [matches, versions, images] = data.reduce(
+        ([m, v, i], d) => {
           m[d.name] = d.full_size
           v.push(d.name)
-          return [m, v]
+          i[d.name] = d.images
+          return [m, v, i]
         },
-        [{}, []]
+        [{}, [], {}]
       )
+
       const version = latest(versions)
-      return { size: matches[version] }
+
+      Object.keys(images).forEach(ver => {
+        if (ver === version) {
+          Object.values(images[ver]).forEach(img => {
+            if (img.architecture === arch) {
+              sizeFromSpecifiedArchitecture = img.size
+            }
+          })
+        }
+      })
+
+      return { size: sizeFromSpecifiedArchitecture || matches[version] }
     } else {
-      return { size: data.full_size }
+      // If the tag is specified, check if any of the returned images has an architecture matching the arch parameter supplied by the user.
+      // If yes, return the size of the image with this arch.
+      // If not, return the value of the `full_size` from the response (the image with the `latest` tag).
```

Comment:
I think rather than falling back to `full_size`, we should throw a `NotFound()` if the `arch` is not found. We should only use `full_size` if no arch is specified. We can also make sure the `arch` value itself is valid and throw a sensible error before we start using this approach https://github.com/badges/shields/blob/3aae8ced784c6b3ab5c6490597749396346e3db0/services/docker/docker-version.service.js#L32-L43 (but we shouldn't default to `amd64` given we don't actually need an `arch` for this badge).

Imagine you are a user trying to make a badge and specifying the `arch` param wrong, an informative error is way more useful than just reporting a number that isn't the number you wanted.

Reply:
Good point, thanks.

I adjusted the code so that now the full_size is used only if no architecture was specified.
If the user supplied the arch param:
if the image with the given arch was found, we give its size,
if the image with the given arch was not found, we throw an `architecture not found` error.

I moved validArchitectures to docker helper file,
and I added two more possible architectures - `386` as a number and `riscv64` (info from https://golang.org/doc/install/source#environment).